### PR TITLE
ui changes for sprint 5

### PR DESCRIPTION
**Pull request by RichardSmithONS**
Squashed branch with the following changes:
Various cosmetic and layout changes requested
Integrate with API changes so uses formattedAddress, classification and geography coordinates if present on the response.
Improved error handler which can be customised in the conf to turn on and off the custom errors for each mode.
Note that for classification we agreed to output just the code for now (if returned) rather than doing lookup - new task in next  sprint for nested output
Singlematch controller tests now use mocked API

### CODE REVIEW

File: demo-ui/app/uk/gov/ons/addressIndex/demoui/ErrorHandler.scala [hunk 1]
```diff
@@ -1,34 +1,64 @@
...
+import javax.inject.{Inject, Singleton}
+
+import play.api.mvc.Results._
+import play.api.Mode.Mode
+import uk.gov.ons.addressIndex.demoui.modules.DemouiConfigModule
 
 @Singleton
-class ErrorHandler extends HttpErrorHandler {
+class ErrorHandler @Inject() (val messagesApi: MessagesApi,
+                                  environment: Environment,
```
Comment (iRhysBradbury) [inline demo-ui/app/uk/gov/ons/addressIndex/demoui/ErrorHandler.scala]: 2 spaces = 1 indentation

File: demo-ui/app/uk/gov/ons/addressIndex/demoui/ErrorHandler.scala [hunk 2]
```diff
@@ -1,34 +1,64 @@
...
-    BadRequest("Bad Request: " + error)
+  val mode : Mode = environment.mode
+
+  // Boolean to determine whether or not to use custom error processing
+  // Settings for each mode come from application.conf file
+  val processError : Boolean = mode match {
+    case Mode.Dev => conf.config.customErrorDev
+    case Mode.Test => conf.config.customErrorTest
+    case Mode.Prod => conf.config.customErrorProd
+    case _ => false
```
Comment (iRhysBradbury) [inline demo-ui/app/uk/gov/ons/addressIndex/demoui/ErrorHandler.scala]: No need to have a `case _`, `Mode` is a Scala `enum`

File: demo-ui/app/uk/gov/ons/addressIndex/demoui/ErrorHandler.scala [hunk 3]
```diff
@@ -1,34 +1,64 @@
...
+                              ) extends HttpErrorHandler with I18nSupport {
 
   val logger = Logger("ErrorHandler")
 
-  def onBadRequest(request: RequestHeader, error: String) = {
-    logger error s"bad request: $error"
-    BadRequest("Bad Request: " + error)
+  val mode : Mode = environment.mode
+
+  // Boolean to determine whether or not to use custom error processing
```
Comment (iRhysBradbury) [inline demo-ui/app/uk/gov/ons/addressIndex/demoui/ErrorHandler.scala]: maybe a few more lines of scaladoc here please

File: demo-ui/app/uk/gov/ons/addressIndex/demoui/ErrorHandler.scala [hunk 4]
```diff
@@ -1,34 +1,64 @@
...
+    }
   }
 
   def onServerError(request: RequestHeader, exception: Throwable) = {
-    logger error s"server error: ${exception.getMessage}"
-    Logger("onServerError").error(exception.getMessage)
-    Future.successful(
-      InternalServerError
-    )
+    if (processError){
```
Comment (iRhysBradbury) [inline demo-ui/app/uk/gov/ons/addressIndex/demoui/ErrorHandler.scala]: `if (processError){` -> `if (processError) {`

File: demo-ui/app/uk/gov/ons/addressIndex/demoui/ErrorHandler.scala [hunk 5]
```diff
@@ -1,34 +1,64 @@
...
+  // Settings for each mode come from application.conf file
+  val processError : Boolean = mode match {
+    case Mode.Dev => conf.config.customErrorDev
+    case Mode.Test => conf.config.customErrorTest
+    case Mode.Prod => conf.config.customErrorProd
+    case _ => false
+  }
+  Logger.info("custom error = " + processError)
+
+   def onBadRequest(request: RequestHeader, error: String) = {
```
Comment (iRhysBradbury) [inline demo-ui/app/uk/gov/ons/addressIndex/demoui/ErrorHandler.scala]: no return type on public method

File: demo-ui/app/uk/gov/ons/addressIndex/demoui/ErrorHandler.scala [hunk 6]
```diff
@@ -1,34 +1,64 @@
...
+      logger error s"bad request: $error"
+      BadRequest(uk.gov.ons.addressIndex.demoui.views.html.error(400, error))
   }
 
   def onClientError(request: RequestHeader, statusCode: Int, message: String) = {
```
Comment (iRhysBradbury) [inline demo-ui/app/uk/gov/ons/addressIndex/demoui/ErrorHandler.scala]: no return type on public method

File: demo-ui/app/uk/gov/ons/addressIndex/demoui/ErrorHandler.scala [hunk 7]
```diff
@@ -1,34 +1,64 @@
...
+      logger error s"client error: $message"
+      Future.successful(
+        NotFound(uk.gov.ons.addressIndex.demoui.views.html.error(statusCode, message))
+      )
+    } else {
+      defaultHttpErrorHandler.onClientError(request, statusCode, message)
+    }
   }
 
   def onServerError(request: RequestHeader, exception: Throwable) = {
```
Comment (iRhysBradbury) [inline demo-ui/app/uk/gov/ons/addressIndex/demoui/ErrorHandler.scala]: no return type on public method

File: demo-ui/app/uk/gov/ons/addressIndex/demoui/ErrorHandler.scala [hunk 8]
```diff
@@ -1,34 +1,64 @@
...
+                                  conf : DemouiConfigModule,
+                                  defaultHttpErrorHandler: DefaultHttpErrorHandler
+                              ) extends HttpErrorHandler with I18nSupport {
 
   val logger = Logger("ErrorHandler")
 
-  def onBadRequest(request: RequestHeader, error: String) = {
-    logger error s"bad request: $error"
-    BadRequest("Bad Request: " + error)
+  val mode : Mode = environment.mode
```
Comment (iRhysBradbury) [inline demo-ui/app/uk/gov/ons/addressIndex/demoui/ErrorHandler.scala]: `val mode : Mode` -> `val mode: Mode`

File: demo-ui/app/uk/gov/ons/addressIndex/demoui/ErrorHandler.scala [hunk 9]
```diff
@@ -1,34 +1,64 @@
...
+
+  // Boolean to determine whether or not to use custom error processing
+  // Settings for each mode come from application.conf file
+  val processError : Boolean = mode match {
+    case Mode.Dev => conf.config.customErrorDev
+    case Mode.Test => conf.config.customErrorTest
+    case Mode.Prod => conf.config.customErrorProd
+    case _ => false
+  }
+  Logger.info("custom error = " + processError)
```
Comment (iRhysBradbury) [inline demo-ui/app/uk/gov/ons/addressIndex/demoui/ErrorHandler.scala]: use the member `logger`

File: demo-ui/app/uk/gov/ons/addressIndex/demoui/ErrorHandler.scala [hunk 10]
```diff
@@ -1,34 +1,64 @@
...
 
   def onServerError(request: RequestHeader, exception: Throwable) = {
-    logger error s"server error: ${exception.getMessage}"
-    Logger("onServerError").error(exception.getMessage)
-    Future.successful(
-      InternalServerError
-    )
+    if (processError){
+      logger error s"server error: ${exception.getMessage}"
+      Logger("onServerError").error(exception.getMessage)
```
Comment (iRhysBradbury) [inline demo-ui/app/uk/gov/ons/addressIndex/demoui/ErrorHandler.scala]: use the member `logger`

File: demo-ui/app/uk/gov/ons/addressIndex/demoui/controllers/SingleMatchController.scala [hunk 11]
```diff
@@ -65,7 +65,7 @@ class SingleMatchController @Inject()(
           Ok(viewToRender)
       )
     } else {
-      Logger.info("Single Match with supplied input address" + addressText)
+      Logger.info("Single Match with supplied input address " + addressText)
```
Comment (iRhysBradbury) [inline demo-ui/app/uk/gov/ons/addressIndex/demoui/controllers/SingleMatchController.scala]: use a member `logger`

File: model/src/main/scala/uk/gov/ons/addressIndex/model/config/AddressIndexConfig.scala [hunk 12]
```diff
@@ -39,13 +39,17 @@ case class IndexesConfig(
 )
 
 case class DemouiConfig (
-   defaultLanguage: String,
-   apiURL: String
+  customErrorDev: Boolean,
+  customErrorTest: Boolean,
+  customErrorProd: Boolean,
+  apiURL: String
```
Comment (iRhysBradbury) [inline model/src/main/scala/uk/gov/ons/addressIndex/model/config/AddressIndexConfig.scala]: `api: ApiConfig`

```
case class(
  host: String,
  port: Int
)
```

Change `AddressIndexClientInstance` to interpolate appropriately

File: demo-ui/test/uk/gov/ons/addressIndex/demoui/controllers/SingleMatchTest.scala [hunk 13]
```diff
@@ -55,23 +56,27 @@ class SingleMatchTest extends PlaySpec with Results {
...
       val configuration = app.injector.instanceOf[DemouiConfigModule]
-      val apiClient = app.injector.instanceOf[AddressIndexClientInstance]
-      val expectedString = "<h3 class=\"green\">10 addresses found</h3>"
+      val apiClient = app.injector.instanceOf[AddressIndexClientMock]
+      val expectedString = "<h3 class=\"green\">1 addresses found</h3>"
 
       // When
-      val response = new SingleMatchController(configuration, messagesApi, apiClient).doMatchWithInput("7 EX2 6GA").apply(FakeRequest())
+      val response = new SingleMatchController(configuration,
+                                               messagesApi,
```
Comment (iRhysBradbury) [inline demo-ui/test/uk/gov/ons/addressIndex/demoui/controllers/SingleMatchTest.scala]: 2 spaces = 1 indentation

File: demo-ui/test/uk/gov/ons/addressIndex/demoui/controllers/SingleMatchTest.scala [hunk 14]
```diff
@@ -55,23 +56,27 @@ class SingleMatchTest extends PlaySpec with Results {
...
+                                               apiClient.asInstanceOf[AddressIndexClientInstance])
+        .doMatchWithInput("7 EX2 6GA").apply(FakeRequest())
```
Comment (iRhysBradbury) [inline demo-ui/test/uk/gov/ons/addressIndex/demoui/controllers/SingleMatchTest.scala]: define test inputs in `given` block

**Comment by RichardSmithONS** [comment]
PR changes applied - codacy was stuck this commit may reset it

**Pull request merged**